cabal-install/src/Distribution/Client/IndexUtils.hs

```diff
@@ -716,15 +720,15 @@ withIndexEntries verbosity (RepoIndex _repoCtxt (RepoLocalNoIndex (LocalRepo nam
                     return (CacheGPD gpd contents)
               where
                 cabalPath = prettyShow pkgid ++ ".cabal"
-            Just pkgId -> do
+            Just pkgId -> runMaybeT $ do
                 -- check for the right named .cabal file in the compressed tarball
-                tarGz <- BS.readFile (localDir </> file)
-                let tar = GZip.decompress tarGz
-                    entries = Tar.read tar
-
-                case Tar.foldEntries (readCabalEntry pkgId) Nothing (const Nothing) entries of
+                tarGz <- lift $ BS.readFile (localDir </> file)
+                tar <- MaybeT $ (Just <$> evaluate (GZip.decompress tarGz)) `catch` \ (e :: Zlib.DecompressError) ->
+                  Nothing <$ warn verbosity ("Cannot read " ++ file ++ ": " ++ show e)
+                let entries = Tar.read tar
+                MaybeT $ case Tar.foldEntries (readCabalEntry pkgId) Nothing (const Nothing) entries of
                     Just ce -> return (Just ce)
-                    Nothing -> die' verbosity $ "Cannot read .cabal file inside " ++ file
+                    Nothing -> Nothing <$ warn verbosity ("Cannot read .cabal file inside " ++ file)
 
```

Comment:
Just a suggestion for more legible code here: 

```haskell
Just pkgId -> do 
  tarGz <- BS.readFile $ localDir </> file 
  tarM <- try (gGZip.decompress tarGz) >>= \case 
    Left (e :: ZLib.DecompressError) -> do 
      warn verbosity $ "Cannot read" ++ file ++ ": " ++ show e
      return Nothing
    Right a -> return $ Just a 
  
  case tarM of  
    Nothing -> return Nothing 
    Just tar -> do 
      let entries = Tar.read tar 
      case Tar.foldEntries (readCabalEntry pkgId) Nothing (const Nothing) entries of 
        Nothing -> do 
          warn verbosity $ "Cannot read .cabal file inside " ++ file 
          return Nothing 
        Just ce -> return $ Just ce 
```

Reply:
I would not bother here at all, but let the exception bail up and handle it with `catches`. There is already a top-level `handle` catching `IOException`, that can handle `Zlib.DecompressError` too.

I don't see much value in bare warning when there is broken tarball. People don't read warnings and open issues that things don't work. Better to fail early.